WalletWasabi/Wallets/P2pBlockProvider.cs

```diff
@@ -0,0 +1,341 @@
+﻿using NBitcoin;
+using NBitcoin.Protocol;
+using NBitcoin.Protocol.Behaviors;
+using System;
+using System.Linq;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.BitcoinCore;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+using WalletWasabi.Models;
+using WalletWasabi.Services;
+
+namespace WalletWasabi.Wallets
+{
+	/// <summary>
+	/// P2pBlocksProvider is a blocks provider that provides blocks 
+	/// from bitcoin nodes using the P2P bitcoin protocol.
+	/// </summary>
+	public class P2pBlockProvider : IBlockProvider
+	{
+		private Node _localBitcoinCoreNode = null;
+
+		public P2pBlockProvider(NodesGroup nodes, CoreNode coreNode, WasabiSynchronizer syncer, ServiceConfiguration serviceConfiguration, Network network)
+		{
+			Nodes = nodes;
+			CoreNode = coreNode;
+			Synchronizer = syncer;
+			ServiceConfiguration = serviceConfiguration;
+			Network = network;
+		}
+
+		public static event EventHandler<bool> DownloadingBlockChanged;
+		public NodesGroup Nodes { get; }
+		public CoreNode CoreNode { get; }
+		public WasabiSynchronizer Synchronizer { get; }
+		public ServiceConfiguration ServiceConfiguration { get; }
+		public Network Network { get; }
+
+		public Node LocalBitcoinCoreNode
+		{
+			get
+			{
+				if (Network == Network.RegTest)
+				{
+					return Nodes.ConnectedNodes.First();
+				}
+
+				return _localBitcoinCoreNode;
+			}
+			private set => _localBitcoinCoreNode = value;
+		}
+
+		private int NodeTimeouts { get; set; }
+
+		/// <summary>
+		/// Gets a bitcoin block from bitcoin nodes using the p2p bitcoin protocol.
+		/// If a CoreNode is available it fetches the blocks using the rpc interface.
```

Comment:
It fetches blocks only from `CoreNodes`? what if it the node is not Core like knots?

Reply:
No, the class is called `CoreNode` and that's why the comment says that. It should be "If a bitcoin node is available..."